design/baremetal-operator/bmh_non-bootable_iso.md

```diff
@@ -0,0 +1,138 @@
+
+## Motivation
+
+In certain scenarios it is desirable to attach a non-bootable iso image
+to a system after provisioning has finished, for example :
+
+A configuration ISO can be used to provide first-boot configuration to a host
+deployed from a standard image in advance. An operator may have a pool of
+powered off hosts with operating system already provisioned. Then, adding
+a host to the cluster boils down to powering it on with a configuration ISO provided.
+
+### Goals
+
+Expose the Ironic API to attach non-bootable iso images via Metal3.
+
+### Non-Goals
+
+Supporting non-ISO images i.e. media types other than CD/DVD, supporting
+drivers other than Redfish and its derived drivers.
+
+## Proposal
+
+### BMO Proposal
+
+Add a new Custom Resource(CR), `DataImage` (with same `name` and `namespace` as the BMH), which contains the spec containing details of the
+non-bootable iso image, example :
+
+#### Design spec
+
+```yaml
+apiVersion: metal3.io/v1alpha1
+kind: DataImage
+metadata:
+  name: metal-worker-0
+  namespace: metal-cluster
+spec:
+  dataImage:
+    url: http://1.2.3.4/image.iso
+```
+
+We will use ownerReferences in the `DataImage` CR to link it to the corresponding `BareMetalHost` object.
+
+Since we are using a CR, a user can create RBAC policies that doesn't allow unauthorized users
+to attach a non-bootable iso image to hosts.
+
+We may consider adding support for HTTP credentials and TLS settings in the future
+but it's outside the scope of this proposal.
+
+**Note** : As part of this proposal, the dataImage will be attached after the node
+has been Provisioned or ExternallyProvisioned.
+
+### Ironic Support
+
+The [proposal](https://bugs.launchpad.net/ironic/+bug/2033288) for exposing this functionaility in Ironic has been
+accepted by the upstream Ironic community.
+
+The [Ironic implementation](https://review.opendev.org/c/openstack/ironic/+/894918) based on the above proposal has been completed.
+
+Other than that, we just need the implementation for driver specific support,
+which in this proposal is Redfish.
+
+## Design Details
+
+Create a new Custom Resource(CR), `DataImage`, with spec field `Url` and same `Name`
+and `Namespace` as the `BareMetalHost` where we are attaching the image.
+
+The ISO will be attached when the `DataImage` CR object is created and the
+corresponding ownerReference is added for the `BareMetalHost` object,  which will trigger a
+reconciliation. In case a reboot is requested at the same time using
+`RebootAnnotation`, the ISO will be attached first.
+
+The `BareMetalHostStatus` will cache the image details including the attachment
+status, which will inform us if the attachment succeeded and can also be used
+for detachment - either when the attachment fails and we retry or when the user
+explicitly requests detachment by deleting the `DataImage` CRD that corresponds to a BMH.
```

Comment:
Either works, but I think we will want the Status thing anyway.

Reply:
I have updated the proposal to add a finalizer to dataImage and also cache it in status.

Reply:
I'm not sure what else we would use BMH.Status for. Just to record the fact that a DataImage exists? I assume we can set up the reconciler so that it gets notified when a DataImage is created with the same name as an existing BMH?

Reply:
UPD: yeah, tracking success is probably a strong enough reason. I just don't really want to update BMH status for this work... Maybe it's enough to use DataImage.Status + a finalizer?

Reply:
We will certainly need to record in the Status of some CR which image we've attached, whether it worked, and if it didn't how many failures we've had.

Note that the only thing that can be updated atomically with the BMH Status is other things in the BMH Status. So to record this stuff in the DataImage Status instead, we'd need to be completely sure that there's no sequence of e.g. API write errors that could cause us to get stuck in an inconsistent state. Are we?

Reply:
Forgot to say: I don't believe we need to update anything in the BMH status in a lockstep?

Reply:
I've updated the proposal to use finalizer + dataImage.status and the BMH reconciler can keep track of the dataImage since it will have the same name and namespace.